journal/src/main/scala/com/evolutiongaming/kafka/journal/util/Pool.scala

```diff
@@ -0,0 +1,77 @@
+package com.evolutiongaming.kafka.journal.util
+
+import cats.effect.kernel.Ref
+import cats.effect.std.Queue
+import cats.effect.{GenConcurrent, Resource}
+import cats.syntax.all._
+import com.evolutiongaming.catshelper.Log
+
+private[journal] trait Pool[F[_], O] {
+
+  /**
+   * `acquire` function of the Resource semantically blocks if there are no objects available.
+   *
+   * `release` function of the Resource returns the object to the pool.
+   * Note that it will NOT deallocate the object.
+   */
+  def borrow: Resource[F, O]
+}
+
+private[journal] object Pool {
+
+  case class ObjectRef[F[_], O](obj: O, deallocate: F[Unit])
+
+  /**
+   * Objects are allocated when a pool is created and deallocated when it shuts down.
+   */
+  def fixedSize[F[_], O](objectOf: Resource[F, O], poolSize: Int, log: Log[F])(
+    implicit F: GenConcurrent[F, Throwable],
+  ): Resource[F, Pool[F, O]] =
+    Resource.make {
+      for {
+        unusedRefs <- Queue.bounded[F, ObjectRef[F, O]](poolSize)
+        usedRefs <- Ref.of[F, Set[ObjectRef[F, O]]](Set.empty)
+        _ <- (1 to poolSize).toVector.traverse_ { _ =>
+          objectOf.allocated.flatMap { case (obj, deallocate) =>
+            unusedRefs.offer(ObjectRef(obj, deallocate))
+          }
+        }
+      } yield new FixedSizePool[F, O](unusedRefs, usedRefs, log)
+    } {
+      _.shutdown
+    }
+
+  private[this] class FixedSizePool[F[_], O](
+    unused: Queue[F, ObjectRef[F, O]],
+    used: Ref[F, Set[ObjectRef[F, O]]],
+    log: Log[F],
+  )(
+    implicit F: GenConcurrent[F, Throwable],
+  ) extends Pool[F, O] {
+
+    override def borrow: Resource[F, O] =
+      Resource.make(acquire)(release).map(_.obj)
+
+    def acquire: F[ObjectRef[F, O]] =
+      for {
+        ref <- unused.take
+        _ <- used.update(_ + ref)
```

Comment:
I'd suggest in order to guarantee that it works 100% to use identity equals and hashcode (default object ones) in ObjectRef. Otherwise it depends on identity of objects inside it (that's how case class equals and hashcode works). The easiest way to do it should be making ObjectRef just a plain final class with `val`s. Don't forget to leave a comment there that it's a plain class to leverage the identity equality of Java Object and use it in sets.

Reply:
> Otherwise it depends on identity of objects inside it
Yes, why do you think it's bad? User code always returns the same object (pointing to the same address in memory) that it got from the pool.

Reply:
I guess this is not relevant anymore since I changed pool implementation to the one you proposed?